File: src/aistudio_client/stream.py
"""Decoder for the newline-delimited positional GenerateContent stream."""

from __future__ import annotations

import json
from collections.abc import Callable
from typing import Any

from .errors import ClientError
from .models import GenerateResult


def visible_text_from_chunk(chunk: Any) -> str:
    if isinstance(chunk, str):
        return chunk
    if isinstance(chunk, dict):
        return str(chunk.get("text") or chunk.get("delta") or "")
    if not isinstance(chunk, list) or not chunk or not isinstance(chunk[0], list):
        return ""
    text = ""
    for frame in chunk[0]:
        try:
            content = frame[0][0][0]
            if not isinstance(content, list) or content[1] != "model":
                continue
            for part in content[0]:
                if isinstance(part, list) and len(part) > 1 and isinstance(part[1], str):
                    text += part[1]
        except (IndexError, TypeError):
            continue
    return text


def collect_generate_result(response, on_chunk: Callable[[Any], None] | None = None) -> GenerateResult:
    result = GenerateResult()
    for raw_line in response.iter_lines(decode_unicode=True):
        if not raw_line:
            continue
        try:
            chunk = json.loads(raw_line)
        except json.JSONDecodeError:
            chunk = raw_line
        result.chunks.append(chunk)
        result.final_text += visible_text_from_chunk(chunk)
        if isinstance(chunk, dict):
            result.finish_reason = chunk.get("finishReason", result.finish_reason)
            result.usage = chunk.get("usage", result.usage)
            result.conversation_metadata = chunk.get("conversationMetadata", result.conversation_metadata)
        if on_chunk:
            on_chunk(chunk)
    if not result.chunks and response.status_code == 200:
        raise ClientError("GenerateContent returned an empty stream", phase="STREAM")
    return result

File: internal/aistudio/stream.go
package aistudio

import (
	"bufio"
	"encoding/json"
	"io"
)

func CollectGenerateResult(reader io.Reader, onChunk func(any)) (GenerateResult, error) {
	result := GenerateResult{}
	scanner := bufio.NewScanner(reader)
	scanner.Buffer(make([]byte, 64*1024), 16*1024*1024)
	for scanner.Scan() {
		line := scanner.Bytes()
		if len(line) == 0 {
			continue
		}
		var chunk any
		if err := json.Unmarshal(line, &chunk); err != nil {
			chunk = string(line)
		}
		result.Chunks = append(result.Chunks, chunk)
		result.FinalText += VisibleTextFromChunk(chunk)
		result.ModelParts = append(result.ModelParts, ModelPartsFromChunk(chunk)...)
		if object, ok := chunk.(map[string]any); ok {
			if value := object["finishReason"]; value != nil {
				result.FinishReason = value
			}
			if value := object["usage"]; value != nil {
				result.Usage = value
			}
			if value := object["conversationMetadata"]; value != nil {
				result.ConversationMetadata = value
			}
		}
		if onChunk != nil {
			onChunk(chunk)
		}
	}
	if err := scanner.Err(); err != nil {
		return result, err
	}
	if len(result.Chunks) == 0 {
		return result, NewError("STREAM", "GenerateContent returned an empty stream")
	}
	return result, nil
}

func VisibleTextFromChunk(chunk any) string {
	if text, ok := chunk.(string); ok {
		return text
	}
	if object, ok := chunk.(map[string]any); ok {
		if value, ok := object["text"].(string); ok {
			return value
		}
		if value, ok := object["delta"].(string); ok {
			return value
		}
	}
	text := ""
	for _, frame := range framesFromChunk(chunk) {
		content, ok := modelContent(frame)
		if !ok {
			continue
		}
		for _, raw := range content {
			part, ok := raw.([]any)
			if !ok || len(part) < 2 {
				continue
			}
			if value, ok := part[1].(string); ok {
				text += value
			}
		}
	}
	return text
}

func ModelPartsFromChunk(chunk any) []any {
	result := []any{}
	for _, frame := range framesFromChunk(chunk) {
		content, ok := modelContent(frame)
		if !ok {
			continue
		}
		for _, raw := range content {
			if part, ok := raw.([]any); ok {
				if decoded := DecodePart(part); decoded != nil {
					result = append(result, decoded)
				}
			}
		}
	}
	return result
}

func framesFromChunk(chunk any) []any {
	root, ok := chunk.([]any)
	if !ok || len(root) == 0 {
		return nil
	}
	frames, _ := root[0].([]any)
	return frames
}
func modelContent(frame any) ([]any, bool) {
	first, ok := frame.([]any)
	if !ok || len(first) == 0 {
		return nil, false
	}
	second, ok := first[0].([]any)
	if !ok || len(second) == 0 {
		return nil, false
	}
	third, ok := second[0].([]any)
	if !ok || len(third) == 0 {
		return nil, false
	}
	content, ok := third[0].([]any)
	if !ok || len(content) < 2 {
		return nil, false
	}
	role, _ := content[1].(string)
	parts, _ := content[0].([]any)
	return parts, role == "model"
}
func DecodePart(part []any) any {
	if len(part) > 1 {
		if text, ok := part[1].(string); ok {
			return map[string]any{"text": text}
		}
	}
	if len(part) > 5 {
		if files, ok := part[5].([]any); ok && len(files) > 0 {
			return map[string]any{"fileData": map[string]any{"fileId": files[0]}}
		}
	}
	return nil
}
